README
NAME
    Lexical::Accessor - true private attributes for Moose/Moo/Mouse

SYNOPSIS
       my $accessor = lexical_has identifier => (
          is       => 'rw',
          isa      => Int,
          default  => sub { 0 },
       );
   
       # or...
       lexical_has identifier => (
          is       => 'rw',
          isa      => Int,
          default  => sub { 0 },
          accessor => \$accessor,
       );
   
       # later...
       say $self->$accessor;     # says 0
       $self->$accessor( 1 );    # setter
       say $self->$accessor;     # says 1

DESCRIPTION
    Lexical::Accessor generates coderefs which can be used as methods to
    access private attributes for objects.

    The private attributes are stored inside-out, and do not add any accessors
    to the class' namespace, so are completely invisible to any outside code,
    including any subclasses. This gives your attribute complete privacy:
    subclasses can define a private (or even public) attribute with the same
    name as your private one and they will not interfere with each other.

    Private attributes can not be initialized by Moose/Moo/Mouse constructors,
    but you can safely initialize them inside a `BUILD` sub.

  Functions
    `lexical_has $name?, %options`
        This module exports a function lexical_has which acts much like
        Moose's `has` function, but sets up a private (lexical) attribute
        instead of a public one.

        Because lexical attributes are stored inside-out, the $name is
        completely optional; however a name is recommended because it allows
        better error messages to be generated.

        The lexical_has function supports the following options:

        `is`
            Moose/Mouse/Moo-style `ro`, `rw`, `rwp` and `lazy` values are
            supported. These control what sort of coderef is returned by the
            `lexical_has` function itself.

               my $reader            = lexical_has "foo" => (is => "ro");
               my $accessor          = lexical_has "foo" => (is => "rw");
               my ($reader, $writer) = lexical_has "foo" => (is => "rwp");

            If generating more than one method it is probably clearer to pass
            in scalar references to the `reader`, `writer`, etc methods,
            rather than relying on the return value of the `lexical_has`
            function.

        `reader`, `writer`, `accessor`, `predicate`, `clearer`
            These accept scalar references. The relevant coderefs will be
            plonked into them:

               my ($get_foo, $set_foo);
   
               lexical_has foo => (
                  reader      => \$get_foo,
                  writer      => \$set_foo,
               );

        `default`, `builder`, `lazy`
            Lazy defaults and builders are allowed. Eager (non-lazy) defaults
            and builders are currently disallowed. (Use a `BUILD` sub to set
            private attribute values at object construction time.)

            The default may be either a non-reference value, or a coderef
            which will be called as a method to return the value.

            Builders probably make less sense than defaults because they
            require a method in the class' namespace. The builder may be a
            method name, or the special value '1' which will be interpreted as
            meaning the attribute name prefixed by "_build_". If a coderef is
            provided, this is automatically installed into the class'
            namespace with the "_build_" prefix. (This last feature requires
            Sub::Name.)

        `isa`
            A type constraint for the attribute. Moo-style coderefs are
            accepted (including those generated by MooX::Types::MooseLike), as
            are Moose::Meta::TypeConstraint/MooseX::Types objects, and
            Mouse::Meta::TypeConstraint/MouseX::Types objects, and of course
            Type::Tiny type constraints.

            String type constraints may also be accepted, but only if
            Type::Utils is installed. (String type constraints are reified
            using `dwim_type`.)

        `does`
            As an alternative to `isa`, you can provide a role name in the
            `does` option.

        `coerce`
            A coderef or Type::Coercion object is accepted.

            If the special value '1' is provided, the type constraint object
            is consulted to find the coercion. (This doesn't work for coderef
            type constraints.)

        `trigger`
            A method name or coderef to trigger when a new value is set.

        `auto_deref`
            Boolean indicating whether to automatically dereference array and
            hash values if called in list context.

        `init_arg`
            Must be `undef` if provided at all.

        `required`
            Must be false if provided at all.

        `weak_ref`
            Boolean. Makes the setter weaken any references it is called with.

        `handles`
            Delegates methods. Has slightly different syntax to Moose's option
            of the same name - is required to be an arrayref of pairs such
            that each pair is a scalar ref followed by a method name, a
            coderef, or an arrayref (where the first element is a method name
            or coderef and subsequent elements are curried arguments).

               my ($get, $post);
  
               lexical_has ua => (
                  isa      => 'HTTP::Tiny',
                  default  => sub { 'HTTP::Tiny'->new },
                  handles  => [
                     \$get   => 'get',
                     \$post  => 'post',
                  ],
               );
   
               # later...
               my $response = $self->$get('http://example.net/');

        `initializer`, `traits`, `lazy_build`
            Not currently implemented. Providing any of these options throws
            an error.

        `documentation`, `definition_context`
            Don't do anything, but are allowed; effectively inline comments.

    `HAS_SUB_NAME`
        Indicates the availability of Sub::Name which Lexical::Accessor will
        use if available to name coderefs for the benefit of stack traces.

        It's also used to name any `builder` subs which are automatically
        installed into your class' namespace.

        This function is not exported.

  Class Methods
    `lexical_has`
        This function may also be called as a class method.

  Comparison (benchmarking, etc)
    Lexical::Accessor is almost three times faster than
    MooX::PrivateAttributes, and almost twenty time faster than
    MooseX::Privacy. I'd also argue that it's a more "correct" implementation
    of private accessors as (short of performing impressive PadWalker
    manipulations), the accessors generated by this module are completely
    invisible to subclasses, method dispatch, etc.

    Compared to the usual Moose convention of using a leading underscore to
    indicate a private method (which is a very loose convention; it is quite
    common for subclasses to override such methods!), Lexical::Accessor
    clearly offers much better method privacy. There should be little
    performance hit from using lexical accessors compared to normal Moose
    accessors. (However they are nowhere near the speed of the XS-powered
    accessors that Moo *sometimes* uses and Mouse *usually* uses.)

    See also: `examples/benchmark.pl` bundled with this release.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Lexical-Accessor>.

SUPPORT
    IRC: support is available through in the *#moops* channel on irc.perl.org
    <http://www.irc.perl.org/channels.html>.

SEE ALSO
    MooX::PrivateAttributes, MooX::ProtectedAttributes, MooseX::Privacy,
    Sub::Private, Method::Lexical, etc...

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013-2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Lexical-Accessor
================

Created:      2013-10-21
Home page:    <https://metacpan.org/release/Lexical-Accessor>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Lexical-Accessor>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.007	2014-03-29

 - Much cleaned up Sub::Accessor::Small.

0.006	2014-03-29

 - Split out the most reusable parts of Lexical::Accessor into
   Sub::Accessor::Small.

0.005	2014-03-28

 - Make Lexical::Accessor more even subclassable.

0.004	2014-03-28

 - Make Lexical::Accessor more subclassable.

0.003	2013-12-15

 - If Sub::Name is available, use it to name the exported `lexical_has`
   keyword.

0.002	2013-10-22

 [ Bug Fixes ]
 - Fix is=>'lazy'.

 [ Documentation ]
 - Bundle a benchmarking script as an example.
 - Comparisons with MooseX::Privacy, MooX::PrivateAttributes, etc.
 - Fix minor (yet confusing) typo.

 [ Other ]
 - Default to is=>'rw' instead of is=>'bare'.
 - Use Hash::FieldHash (it's significantly faster) instead of
   Hash::Util::FieldHash::Compat.

0.001	2013-10-21	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "true private attributes for Moose/Moo/Mouse",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.015, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Lexical-Accessor",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       recommends => { "Sub::Name" => 0, "Type::Utils" => 0.026 },
                                       requires   => {
                                                       "Eval::TypeTiny" => 0,
                                                       "Exporter::Tiny" => 0,
                                                       "Hash::FieldHash" => 0,
                                                       "perl" => 5.008003,
                                                     },
                                     },
                        test      => {
                                       recommends => {
                                                       "Class::Tiny" => 0.006,
                                                       "Moo" => "1.000000",
                                                       "Moose" => "2.0000",
                                                       "Mouse" => "1.00",
                                                     },
                                       requires   => {
                                                       "Test::Fatal"     => 0,
                                                       "Test::More"      => 0.96,
                                                       "Test::Requires"  => 0,
                                                       "Types::Standard" => 0.004,
                                                     },
                                     },
                      },
  "provides"       => {
                        "Lexical::Accessor"    => { file => "lib/Lexical/Accessor.pm", version => 0.007 },
                        "Sub::Accessor::Small" => { file => "lib/Sub/Accessor/Small.pm", version => 0.007 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=Lexical-Accessor",
                                        },
                        homepage     => "https://metacpan.org/release/Lexical-Accessor",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-lexical-accessor.git",
                                          web  => "https://github.com/tobyink/p5-lexical-accessor",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/Lexical-Accessor/project",
                      },
  "version"        => 0.007,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

